### PR TITLE
Several updates to the OFI API

### PR DESCRIPTION
**Pull request by shefty**
This pull request is NOT ready for merging.  I'm still debugging fabtests support over one of the changes.

Please see the individual commit messages for details.

2 patches deal with report user data as part of connection establishment.
3 attempt to simplify the fi_getinfo caps fields -- providing clearer documentation and removing 2 caps that are redundant.
1 patch adds the optype to the completion data.

@jithinjosepkl @goodell @j-xiong @a-ilango - Note that you may need to make some minor updates to the providers to ensure that they are compliant.  I did make a few updates, where directed by searches.

### CODE REVIEW

**Comment by sayantansur** [comment]
These look good to me.

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -1744,7 +1744,16 @@ fi_ibv_msg_ep_getopt(fid_t fid, int level, int optname,
 {
 	switch (level) {
 	case FI_OPT_ENDPOINT:
-		return -FI_ENOPROTOOPT;
+		switch (optname) {
+		case FI_OPT_CM_DATA_SIZE:
+			if (*optlen < sizeof(size_t))
+				return -FI_ETOOSMALL;
+			*((size_t *) optval) = 56;
```
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: How was this size, 56 chosen? Can we use a macro for this?

Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: I had to read the IB spec, calculate the size of the RDMA CM header, and this is the size that was left.  :P

**Comment by j-xiong** [comment]
Looks good to me. I will update the PSM provider accordingly. One exception maybe the reference of FI_REMOTE_CQ_DATA. they need to be removed to pass compilation.

**Comment by shefty** [comment]
The FI_REMOTE_CQ_DATA flag still remains.  It's needed as an operational flag and on the completion side.  It's only removed as a capability bit.

**Comment by jithinjosepkl** [comment]
Sockets provider changes also look good.
(tagging @shantonu)

**Comment by goodell** [comment]
I'll update the usnic provider some time this week.  @shefty, do you need my updates before you merge this PR?

**Comment by shefty** [comment]
I don't need any updates before merging this.  I'll merge as soon as I can find why the changes to FI_INJECT break the sample app.

**Pull request merged**